docs/source/generics.rst

```diff
@@ -0,0 +1,984 @@
+         # The rotate method is only defined on Circle, not on Shape
+         def rotate(self): ...
+
+     def add_one(things: list[Shape]) -> None:
+         things.append(Shape())
+
+     my_circles: list[Circle] = []
+     add_one(my_circles)     # This may appear safe, but...
+     my_circles[-1].rotate()  # ...this will fail, since my_circles[0] is now a Shape, not a Circle
+
+  Another example of invariant type is :py:class:`~typing.Dict`. Most mutable containers
+  are invariant.
+
+By default, all user-defined generics are invariant.
+To declare a given generic class as covariant or contravariant use
+type variables defined with special keyword arguments ``covariant`` or
+``contravariant``. For example:
+
+.. code-block:: python
+
+   from typing import Generic, TypeVar
+
+   T_co = TypeVar('T_co', covariant=True)
+
+   class Box(Generic[T_co]):  # this type is declared covariant
+       def __init__(self, content: T_co) -> None:
+           self._content = content
+
+       def get_content(self) -> T_co:
+           return self._content
+
+   def look_into(box: Box[Animal]): ...
+
+   my_box = Box(Cat())
+   look_into(my_box)  # OK, but would be an error if Box was invariant in T
+
+.. _type-variable-upper-bound:
+
+Type variables with upper bounds
+********************************
+
+By default, a type variable can be replaced with any type. This means that
+you can't do very much with an object of type ``T`` safely -- you don't
+know anything about it!
+
+It's therefore often useful to be able to limit the types that a type
+variable can take on, for instance, by restricting it to values that are
+subtypes of a specific type.
+
+Such a type is called the upper bound of the type variable, and is specified
+with the ``bound=...`` keyword argument to :py:class:`~typing.TypeVar`.
+
+.. code-block:: python
+
+    from typing import TypeVar, SupportsAbs
+
+    T = TypeVar('T', bound=SupportsAbs[float])
+
+In the definition of a generic function that uses such a type variable
+``T``, the type represented by ``T`` is assumed to be a subtype of
+its upper bound, so the function can use methods of the upper bound on
+values of type ``T``.
+
+.. code-block:: python
+
+    def largest_in_absolute_value(*xs: T) -> T:
+        return max(xs, key=abs)  # Okay, because T is a subtype of SupportsAbs[float].
+
+In a call to such a function, the type ``T`` must be replaced by a
+type that is a subtype of its upper bound. Continuing the example
+above:
+
+.. code-block:: python
+
+    largest_in_absolute_value(-3.5, 2)   # OK, has type float
+    largest_in_absolute_value(5+6j, 7)   # OK, has type complex
+    largest_in_absolute_value('a', 'b')  # error: error: Value of type variable "T" of "largest_in_absolute_value" cannot be "str"
+
+Type parameters of generic classes may also have upper bounds, which
+restrict the valid values for the type parameter in the same way.
+
+.. _type-variable-value-restriction:
+
+Type variables with value restriction
```

Comment:
The term "value restriction" is slightly confusing here, because type variable constraints work on types, i.e. they restrict the set of types a type parameter can take.

I think PEP-484 and the `typing` docs use the term "constraints", so it's probably best to use it here as well.

Reply:
Yes, we should use "TypeVar with constraints" for AnyStr-like TypeVars.

Reply:
Thanks, fixed!